### PR TITLE
HDDS-8526. Intermittent failure: Directory is not empty

### PR DESCRIPTION
**Pull request by ashishkumar50**
## What changes were proposed in this pull request?

When key is deleted, it firsts inserts into cache as deleted and later flush to db which deletes entry from db. In case of Bucket empty check we first check in cache and then check in table whether any entry exist for the prefix. But during this iteration there is chance that cache gets flushed into db and when we read cache it returns NULL and at this point we assume that table has entry but cache don't have. This shows that bucket is not empty at this point.
So to fix this in this case we should check in table whether really table entry is removed.

Similar problem is for Volume empty check. Making volume empty check also follow above criteria to ensure it is empty or not.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8526

## How was this patch tested?
Before Fix: Test fails Intermittently in latest master for TestRootedOzoneFileSystemWithFSO.
https://github.com/ashishkumar50/ozone/actions/runs/5529154450

After Fix: Succeeded in all 10 * 10 run for TestRootedOzoneFileSystemWithFSO.
https://github.com/ashishkumar50/ozone/actions/runs/5531565163

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 1]
```diff
@@ -1045,6 +1014,17 @@ private <T> boolean isKeyPresentInTable(String keyPrefix,
...
+          // There is chance cache value flushed when
+          // we iterate through the table.
+          // Check in table whether it is deleted or still present.
+          if (table.getIfExist(kv.getKey()) == null) {
+            if (!keyIter.hasNext()) {
+              break;
+            }
+            kv = keyIter.next();
+            continue;
+          }
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:1026]: Nit: we could avoid duplicating steps for `keyIter` by converting to `if-else`.  Something like this:

```java
        // Case 1: We found an entry, but no cache entry.
        if (cacheValue == null) {
          // we found at least one key with this prefix.
          // There is chance cache value flushed when
          // we iterate through the table.
          // Check in table whether it is deleted or still present.
          if (table.getIfExist(kv.getKey()) != null) {
            // Still in table and no entry in cache
            return true;
          }
        } else if (cacheValue.getCacheValue() != null) {
          // Case 2a:
          // We found a cache entry and cache value is not null.
          return true;
        }
```

**Review by adoroszlai — COMMENTED** [review]
Thanks a lot @ashishkumar50 for fixing this annoying bug.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java (hunk 1 above)
Comment (ashishkumar50) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java]: Thanks for the suggestion, Updated.

**Review by adoroszlai — APPROVED** [review]

**Review by sumitagrawl — COMMENTED** [review]
@ashishkumar50 Thanks for working over this, there is an improvement possible,
BucketTable is cache CacheType.FULL_CACHE, i.e. entire bucket table is cached. So this actually do not need to check the Table over disk.

The bug here is fixing that, if entry in disk table is present and validating with cache entry exist and reverifying with disk. But check with disk table can be avoided.

**Comment by ashishkumar50** [comment]
> @ashishkumar50 Thanks for working over this, there is an improvement possible, BucketTable is cache CacheType.FULL_CACHE, i.e. entire bucket table is cached. So this actually do not need to check the Table over disk.
> 
> The bug here is fixing that, if entry in disk table is present and validating with cache entry exist and reverifying with disk. But check with disk table can be avoided.

@sumitagrawl, You mean for checking isVolumeEmpty we should avoid calling
`isKeyPresentInTable(volumePrefix)` ? And we can just depend on cache?

Old existing code was checking even in DB. Is it safe here just depend on cache and not checking on disk?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 2]
```diff
@@ -1045,15 +1014,18 @@ private <T> boolean isKeyPresentInTable(String keyPrefix,
         // Case 1: We found an entry, but no cache entry.
         if (cacheValue == null) {
           // we found at least one key with this prefix.
+          // There is chance cache value flushed when
+          // we iterate through the table.
+          // Check in table whether it is deleted or still present.
+          if (table.getIfExist(kv.getKey()) != null) {
+            // Still in table and no entry in cache
+            return true;
+          }
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:1023]: Thanks @ashishkumar50 for the patch.

While it looks like the issue is indeed [mitigated](https://github.com/apache/ozone/blob/e8fbdaaca2a32c67aeb5ba8fbd23bb2a271480f2/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketDeleteRequest.java#L136-L141) by this extra check, it sounds bizzare to me that the `bucketTable` with `FULL_CACHE` would remove the cache entry after a disk flush in the first place.

https://github.com/apache/ozone/blob/d1ea589314280e6bdc9380bdbb329bb3ba307bb0/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L637-L646

IIUC, the whole point of a `FULL_CACHE` table is to keep all entries in cache. And all DB entries should have been loaded into cache upon OM startup:

https://github.com/apache/ozone/blob/4b9b10f6206ed5c29c0c4957a651de1d24fe86f6/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java#L110-L125

**Review by smengcl — APPROVED** [review]

**Comment by adoroszlai** [comment]
> recently we have done same for container delete, even though we have in cache that shows container is empty but still we check from db to really ensure it is empty before delete

Only volume and bucket tables use `FULL_CACHE`, so other table may have entries not in the cache.

[100x run passed](https://github.com/adoroszlai/hadoop-ozone/actions/runs/5537988977) with the [change](https://github.com/adoroszlai/hadoop-ozone/commit/dab1c0fac24509df0eb75c13182a34933fd6a980) @sumitagrawl suggested.

Regular CI:
https://github.com/adoroszlai/hadoop-ozone/actions/runs/5540255300

To be on the safe side, we can make the change in two steps (for easier revert if problems surface): first merge this one as is, then create new PR for the additional performance improvement.

**Comment by ashishkumar50** [comment]
> > recently we have done same for container delete, even though we have in cache that shows container is empty but still we check from db to really ensure it is empty before delete
> 
> Only volume and bucket tables use `FULL_CACHE`, so other table may have entries not in the cache.
> 
> [100x run passed](https://github.com/adoroszlai/hadoop-ozone/actions/runs/5537988977) with the [change](https://github.com/adoroszlai/hadoop-ozone/commit/dab1c0fac24509df0eb75c13182a34933fd6a980) @sumitagrawl suggested.
> 
> Regular CI: https://github.com/adoroszlai/hadoop-ozone/actions/runs/5540255300
> 
> To be on the safe side, we can make the change in two steps (for easier revert if problems surface): first merge this one as is, then create new PR for the additional performance improvement.

Ok we can do improvement in separate PR, currently we can keep as old behaviour to check both Cache and DB for bucket and key table. Later we can use just cache for bucket table.

**Comment by adoroszlai** [comment]
Thanks @ashishkumar50 for the fix, @smengcl, @sumitagrawl for the review.

**Pull request merged**